sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -751,6 +753,49 @@ abstract class TypeCoercionBase {
     }
   }
 
+  /**
+   * Casts types according to bucket info for Equality expression.
+   */
+  object EqualityTypeCasts extends TypeCoercionRule {
+
+    private def isIntegralTypes(exprs: Seq[Expression]): Boolean = {
+      exprs.map(_.dataType).forall {
+        case _: IntegralType => true
+        case DecimalType.Fixed(_, 0) => true
+        case _ => false
+      }
+    }
+
+    override val transform: PartialFunction[Expression, Expression] = {
+      case b @ Equality(l: Attribute, r: Attribute)
+          if b.childrenResolved && l.dataType != r.dataType && isIntegralTypes(b.children) =>
+        // The result type is:
+        // 1. The attribute data type with larger bucket number.
+        // 2. The attribute data type with larger default size if it is not bucketed attribute.
```

Comment:
I don't get it. Can you explain more? Are you saying that adding cast to either side is correct, and we need to make a decision with least cost?

Reply:
Yes. adding cast to either side is correct for equality expressions.

Reply:
but this is not true under ansi mode, right?

Reply:
can we improve the bucket scan planning code to accept safe cast?

Reply:
We can use `TRY` mode if under ansi mode:
https://github.com/apache/spark/blob/600dc6cf86bd66d6b92d73cd8dda7ac39a166032/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L785-L791

Reply:
> can we improve the bucket scan planning code to accept safe cast?

Do you mean something like [this changes](https://github.com/apache/spark/pull/32063/files)?

Reply:
Ah yes! Can you reopen and update that PR?

Reply:
one more pr https://github.com/apache/spark/pull/34755 : )

Reply:
That change is incorrect. For example:
```scala
import org.apache.spark.sql.catalyst.expressions.{Cast, EmptyRow, Literal, Murmur3Hash, Pmod}
import org.apache.spark.sql.types.LongType

println(Pmod(new Murmur3Hash(Seq(Literal(100))), Literal(10)).eval(EmptyRow))
println(Pmod(new Murmur3Hash(Seq(new Cast(Literal(100), LongType))), Literal(10)).eval(EmptyRow))
```

Output:
```
3
0
```

Related code:
https://github.com/apache/spark/blob/69aa727ff495f6698fe9b37e952dfaf36f1dd5eb/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L505-L508

Reply:
good point ! the value is equivalent but the hash code can be different according to the data type

Reply:
New PR: https://github.com/apache/spark/pull/40805